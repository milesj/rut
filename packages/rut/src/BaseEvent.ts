export default class BaseEvent<T = unknown> {
  bubbles: boolean = true;

  cancelable: boolean = true;

  currentTarget!: T;

  defaultPrevented: boolean = false;

  eventPhase: number = 0;

  isTrusted: boolean = true;

  propagationStopped: boolean = false;

  srcElement: null = null;

  target!: T;

  timeStamp: number;

  type: string;

  constructor(type: string) {
    this.timeStamp = Date.now();
    this.type = type;
  }

  preventDefault(): void {
    this.defaultPrevented = true;
  }

  stopImmediatePropagation(): void {
    this.propagationStopped = true;
  }

  stopPropagation(): void {
    this.propagationStopped = true;
  }
}
