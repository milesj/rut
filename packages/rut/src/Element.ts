/* eslint-disable @typescript-eslint/no-unused-vars */

import React from 'react';
import { ReactTestInstance } from 'react-test-renderer';
import { getTypeName } from './internals/react';
import { doAct, doAsyncAct } from './internals/act';
import { debug } from './internals/debug';
import { getPropForDispatching } from './internals/element';
import { whereTypeAndProps } from './predicates';
import {
  Predicate,
  DispatchOptions,
  DebugOptions,
  UnknownProps,
  AtIndexType,
  QueryOptions,
  AdapterRendererOptions,
  ElementType,
} from './types';

export default abstract class Element<
  Type extends ElementType = ElementType,
  Props = never,
  Host = unknown
> {
  // @ts-expect-error Set after instantiation
  options: AdapterRendererOptions;

  protected element: ReactTestInstance;

  private readonly isRutElement = true;

  constructor(element: ReactTestInstance) {
    this.element = element;
  }

  /**
   * Log and return a JSX representation of the current *reconciled* React component tree.
   * Does not include exotic components or nodes, such as:
   *
   *  - Context consumers and providers
   *  - Roots, portals, modes
   *  - Profiler, Suspense
   *  - Fragments
   */
  debug = (options: DebugOptions = {}) => debug(this.element, options);

  /**
   * Dispatch an event listener for the defined prop name. Requires a synthetic event
   * based on the original event type.
   *
   * Note: This may only be executed on host components.
   */
  dispatch(name: string, eventOrConfig?: unknown, options: DispatchOptions = {}): this {
    const prop = getPropForDispatching(this, name);
    const event = this.createSyntheticEvent(name, eventOrConfig, this.element.type);

    // istanbul ignore next
    if (options.propagate) {
      // eslint-disable-next-line no-console
      console.warn('Event propagation is experimental and is not fully implemented.');
    }

    doAct(() => prop(event), this.options.applyPatches);

    return this;
  }

  /**
   * Like `dispatch` but also awaits the event so that async calls have time to finish.
   */
  async dispatchAndWait(
    name: string,
    eventOrConfig?: unknown,
    options: DispatchOptions = {},
  ): Promise<void> {
    const prop = getPropForDispatching(this, name);
    const event = this.createSyntheticEvent(name, eventOrConfig, this.element.type);

    // istanbul ignore next
    if (options.propagate) {
      // eslint-disable-next-line no-console
      console.warn('Event propagation is experimental and is not fully implemented.');
    }

    await doAsyncAct(() => prop(event), this.options.applyPatches);
  }

  /**
   * Search through the current tree for all elements that match the defined React
   * component or HTML type. If any are found, a list of `Element`s is returned.
   */
  find(type: ElementType, props?: UnknownProps): Element<ElementType>[] {
    return this.query(whereTypeAndProps(type, props));
  }

  /**
   * Search through the current tree for all elements that match the defined React
   * component or HTML type. If any are found, return the `Element` at the defined
   * index. Accepts shorthand `first` and `last` indices.
   */
  findAt(type: ElementType, at: AtIndexType, props?: UnknownProps): Element<ElementType> {
    const results = this.query(whereTypeAndProps(type, props));
    let index: number;

    if (at === 'first') {
      index = 0;
    } else if (at === 'last') {
      index = results.length - 1;
    } else if (typeof at === 'number') {
      index = at;
    } else {
      throw new TypeError(`Invalid index type "${at}".`);
    }

    const value = results[index];

    if (!value) {
      throw new Error(
        `Expected to find an element at index ${index} for \`${getTypeName(type)}\`.`,
      );
    }

    return value;
  }

  /**
   * Search through the current tree for all elements that match the defined React
   * component or HTML type. If exactly 1 is found, a `Element`s is returned,
   * otherwise an error is thrown.
   */
  findOne(type: ElementType, props?: UnknownProps): Element<ElementType> {
    const results = this.find(type, props);

    if (results.length !== 1) {
      throw new Error(
        `Expected to find 1 element for \`${getTypeName(type)}\`, found ${results.length}.`,
      );
    }

    return results[0];
  }

  /**
   * Return the name of the component, element, or node as JSX.
   */
  name(jsx: boolean = false): string {
    // Use the raw fiber types for names, as they preserve the internal node structures
    const name = getTypeName(this.element._fiber.elementType || this.element._fiber.type);

    return jsx ? `<${name} />` : name;
  }

  /**
   * Search through the current child tree using a custom predicate, which is passed the
   * ReactTestRenderer node and internal React fiber node. If any are found,
   * a list of `Element`s is returned.
   */
  query<T extends ElementType>(predicate: Predicate, options?: QueryOptions): Element<T>[] {
    return this.element
      .findAll((node) => predicate(node, node._fiber), { deep: true, ...options })
      .map((node) => {
        const element = this.options.createElement(node);

        element.options = this.options;

        return element;
      });
  }

  /**
   * Return a ref associated with the current component.
   * If a ref is stored on the internal fiber node, it will be returned.
   * Otherwise an instance, callback, or string ref may be referenced by name.
   */
  ref<T>(name?: string): T | null {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const inst = this.element.instance;
    const fiber = this.element._fiber;

    if (!name || fiber.ref) {
      return fiber.ref;
    }

    // Callback refs
    if (name in inst) {
      return inst[name];
    }

    // String refs
    if (inst.refs[name]) {
      return inst.refs[name];
    }

    return null;
  }

  /**
   * Return name when cast as a string.
   */
  toString(): string {
    return this.name(true);
  }

  /**
   * Create a synthetic event specific to this renderer.
   */
  abstract createSyntheticEvent(
    name: string,
    eventOrConfig: unknown,
    type: ElementType,
  ): React.SyntheticEvent;
}
