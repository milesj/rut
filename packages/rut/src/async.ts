/* eslint-disable @typescript-eslint/no-explicit-any */

const NativePromise = (global.Promise as any) as PromiseConstructor;
const methodsToWrap: (keyof PromiseConstructor)[] = ['all', 'race', 'resolve', 'reject'];

export function wrapAndCaptureAsync(): Promise<any>[] {
  const queue: Promise<any>[] = [];

  function MockPromise<T>(
    executor: (
      resolve: (value?: T | PromiseLike<T>) => void,
      reject: (reason?: any) => void,
    ) => void,
  ): Promise<T> {
    const promise = new NativePromise(executor);

    queue.push(promise);

    return promise;
  }

  methodsToWrap.forEach(method => {
    Object.defineProperty(MockPromise, method, {
      value: (...args: unknown[]) => {
        // @ts-ignore
        const promise = NativePromise[method](...args);

        queue.push(promise);

        return promise;
      },
    });
  });

  global.Promise = MockPromise;

  return queue;
}

export function unwrapPromise() {
  global.Promise = NativePromise;
}

export function waitForAsyncQueue(queue: Promise<any>[]) {
  unwrapPromise();

  return NativePromise.all(queue);
}
