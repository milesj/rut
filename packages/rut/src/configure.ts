import { RendererOptions, IntegrationOptions } from './types';

export const globalOptions: RendererOptions = {};

export function configure(options: RendererOptions) {
  Object.assign(globalOptions, options);
}

export const integrationOptions: IntegrationOptions = {
  runWithTimers(cb) {
    return cb();
  },
};

export function configureIntegration(options: IntegrationOptions) {
  Object.assign(integrationOptions, options);
}
