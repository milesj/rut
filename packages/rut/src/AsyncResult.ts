import Result from './Result';
import { RendererOptions } from './types';
import { doAsyncAct } from './internals/act';
import Element from './Element';

export default class AsyncResult<
  Props extends object = {},
  Root extends Element = Element
> extends Result<Props, Root> {
  /**
   * Like `rerender` but also awaits the re-render so that async calls have time to finish.
   */
  rerenderAndWait = async (element: React.ReactElement<Props>, options?: RendererOptions) => {
    Object.assign(this.options, options);

    await doAsyncAct(
      () =>
        this.renderer.unstable_flushSync(() => this.renderer.update(this.updateElement(element))),
      this.options.applyPatches,
    );

    return this.root;
  };
}
