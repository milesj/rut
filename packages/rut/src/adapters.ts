import React from 'react';
import AsyncResult from './AsyncResult';
import BaseEvent from './BaseEvent';
import Element from './Element';
import { doAct, doAsyncAct } from './internals/act';
import { globalOptions } from './internals/config';
import SyncResult from './SyncResult';
import SyntheticEvent from './SyntheticEvent';
import { AdapterRendererOptions } from './types';

export { BaseEvent, Element, SyntheticEvent };

export function doRender<Props extends object, Root extends Element>(
  element: React.ReactElement<Props>,
  options: AdapterRendererOptions,
): SyncResult<Props, Root> {
  return doAct(
    () =>
      new SyncResult(element, {
        ...globalOptions,
        ...options,
      }),
    options.applyPatches,
  );
}

export async function doRenderAndWait<Props extends object, Root extends Element>(
  element: React.ReactElement<Props>,
  options: AdapterRendererOptions,
): Promise<AsyncResult<Props, Root>> {
  return doAsyncAct(
    () =>
      new AsyncResult(element, {
        ...globalOptions,
        ...options,
      }),
    options.applyPatches,
  );
}
