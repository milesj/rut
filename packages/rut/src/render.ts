import React from 'react';
import AsyncResult from './AsyncResult';
import SyncResult from './SyncResult';
import { globalOptions } from './configure';
import { doAct, doAsyncAct } from './internals/act';
import { RendererOptions } from './types';

// export function render<Props extends object = {}>(
//   element: React.ReactElement<Props>,
//   options?: RendererOptions,
// ): SyncResult<Props> {
//   const opts = {
//     ...globalOptions,
//     ...options,
//   };

//   return doAct(() => new SyncResult(element, opts));
// }

export async function renderAndWait<Props extends object = {}>(
  element: React.ReactElement<Props>,
  options?: RendererOptions,
): Promise<AsyncResult<Props>> {
  const opts = {
    ...globalOptions,
    ...options,
  };

  return doAsyncAct(() => new AsyncResult(element, opts), opts.asyncMode);
}

export const render = renderAndWait;
