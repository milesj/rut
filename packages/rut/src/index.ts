/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { act } from 'react-test-renderer';
import matchers from './matchers';
import { render, renderAndWait } from './render';
import mockSyntheticEvent from './mocks/event';

export * from './helpers';
export * from './types';

export { act, matchers, mockSyntheticEvent, render, renderAndWait };
