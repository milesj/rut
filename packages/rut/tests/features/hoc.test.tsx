import React from 'react';
import { render } from '../../src/render';

describe('HOCs', () => {
  interface WithStylesProps {
    styles: object;
  }

  // Factory style
  function withStyles() {
    return function withStylesFactory<P>(
      WrappedComponent: React.ComponentType<P & WithStylesProps>,
    ) {
      function WithStyles(props: P) {
        return <WrappedComponent {...props} styles={{ color: 'green' }} />;
      }

      WithStyles.displayName = `withStyles(${WrappedComponent.displayName ||
        WrappedComponent.name})`;

      return WithStyles;
    };
  }

  interface WrappedProps {
    id: number;
  }

  class Wrapped extends React.Component<WrappedProps & WithStylesProps> {
    render() {
      return <div>Styled!</div>;
    }
  }

  it('can render the wrapped component + hoc', async () => {
    const StyledComp = withStyles()(Wrapped);
    const result = await render<WrappedProps>(<StyledComp id={1} />);

    expect(result).toMatchSnapshot();
  });

  it('returns the result HOC as the root', async () => {
    const StyledComp = withStyles()(Wrapped);
    const { root } = await render<WrappedProps>(<StyledComp id={2} />);

    expect(root.name()).toBe('withStyles(Wrapped)');
    expect(root).toHaveProps({ id: 2 });
  });

  it('can find the wrapped component', async () => {
    const StyledComp = withStyles()(Wrapped);
    const { root } = await render<WrappedProps>(<StyledComp id={3} />);

    expect(root.find(Wrapped)).toHaveLength(1);
    expect(root.findOne(Wrapped)).toHaveProps({ id: 3, styles: { color: 'green' } });
  });

  interface ConnectProps {
    connected: true;
  }

  // Non-factory style
  function connect<P>(WrappedComponent: React.ComponentType<P & ConnectProps>) {
    function Connect(props: P) {
      return <WrappedComponent {...props} connected />;
    }

    Connect.displayName = `connect(${WrappedComponent.displayName || WrappedComponent.name})`;

    return Connect;
  }

  class DeepWrapped extends React.Component<{ id: number } & WithStylesProps & ConnectProps> {
    render() {
      return <div>Styled!</div>;
    }
  }

  it('supports multiple layers of HOCs', async () => {
    const DeepComp = connect(withStyles()(DeepWrapped));
    const result = await render<WrappedProps>(<DeepComp id={4} />);

    expect(result).toMatchSnapshot();
  });
});
