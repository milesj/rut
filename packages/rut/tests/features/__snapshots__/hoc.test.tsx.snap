// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOCs can render the wrapped component + hoc 1`] = `
<withStyles(Wrapped) id={1}>
  <Wrapped id={1} styles={{ color: "green" }}>
    <div>Styled!</div>
  </Wrapped>
</withStyles(Wrapped)>
`;

exports[`HOCs supports multiple layers of HOCs 1`] = `
<connect(withStyles(DeepWrapped)) id={4}>
  <withStyles(DeepWrapped) connected id={4}>
    <DeepWrapped connected id={4} styles={{ color: "green" }}>
      <div>Styled!</div>
    </DeepWrapped>
  </withStyles(DeepWrapped)>
</connect(withStyles(DeepWrapped))>
`;
